### ISSUE TITLE
Switch our documentation to use `dirhtml` instead of `html`

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/team-compass#666]
Currently we are using the default Sphinx builder for `html` outputs, which creates an individual file with `.html` at the end. For example:

https://compass.2i2c.org/en/latest/open-source/strategy.html

However most modern website builders default to using a **directory** as the build output, with an `index.html` file inside, so that instead the link structure becomes:

https://compass.2i2c.org/en/latest/open-source/strategy

I think this structure is much more common in general these days, and most other SSGs use this structure (including 2i2c.org). It is also easily do-able with Sphinx, via `-b dirhtml`.

I propose that we migrate all of our documentation to use the `dirhtml` builder and update links accordingly.

### Sites we'd need to update

- docs.2i2c.org
- infrastructure.2i2c.org
- compass.2i2c.org

**Comment by consideRatio** [comment]
Do you know if this change break existing links in some way? For example, will https://compass.2i2c.org/en/latest/open-source/strategy.html stop working when https://compass.2i2c.org/en/latest/open-source/strategy starts working?

**Comment by choldgraf** [comment]
yep it will, I think we'll need to do a `ctrl+f` for any `.html` links in our docs and remove it. That shouldn't be too hard as long as it's linked from within `2i2c-org`. E.g. here's a search we can use:

https://github.com/search?q=org%3A2i2c-org+%22.html%29%22&type=code

For any document / label linked directly by Sphinx, it will automatically change, we don't need to change that.

**Comment by consideRatio** [comment]
The key question in my head then is if we accept breaking all external existing links to accomplish this, and/or if we can use rediraffe to help redirect the `something.html` versions to `something/` instead.

**Comment by choldgraf** [comment]
I don't think that rediraffe would work, because it operates at the document level rather than at the end-link level unfortunately :-/

**Comment by damianavila** [comment]
The proposed migration makes sense to me.

**Comment by choldgraf** [comment]
### We could manually define the pages to re-direct

I briefly looked into this - `sphinx-rediraffe` doesn't let you redirect from something like `mypage.html` to `mypage/index.html`. However, the template they are using is pretty simple:

https://github.com/wpilibsuite/sphinxext-rediraffe/blob/d8961c2402015dafa6039ed27d973ad835d7fc09/sphinxext/rediraffe.py#L21-L39

<details><summary>Their template</summary>
<p>

```
DEFAULT_REDIRAFFE_TEMPLATE = Template(
    """
<html>
    <head>
        <noscript>
            <meta http-equiv="refresh" content="0; url={{rel_url}}"/>
        </noscript>
    </head>
    <body>
        <script>
            window.location.href = '{{rel_url}}' + (window.location.search || '') + (window.location.hash || '');
        </script>
        <p>You should have been redirected.</p>
        <a href="{{rel_url}}">If not, click here to continue.</a>
    </body>
</html>
"""
)
```

</p>
</details> 

Then they write the redirect HTML with this

```
        with build_redirect_from.open("w") as f:
            f.write(
                rediraffe_template.render(
                    rel_url=str(
                        PurePosixPath(
                            PureWindowsPath(
                                relpath(build_redirect_to, build_redirect_from.parent)
                            )
                        )
                    ),
                    from_file=src_redirect_from,
                    to_file=src_redirect_to,
                    from_url=redirect_from,
                    to_url=redirect_to,
                )
            )
```

So one option for us would be to write a little Sphinx hook in our sphinx-2i2c-theme that:

1. Checked if `dirhtml` was the active builder
2. If so, then for every page, use the pattern above to manually write HTML that makes every page redirect from `folder/page1.html` to `folder/page1/index.html`

That way, we should be able to do all the redirects without needing to make per-theme changes. I don't _think_ it would be too much work, as long as we weren't worried about some of the edge cases (which I don't think we should be as this is just a stop-gap).

### If you don't think this is a good idea please say so

I'll plan on us moving forward with this in the next few days unless somebody objects.

**Comment by consideRatio** [comment]
Great solution @choldgraf!!

**Comment by choldgraf** [comment]
OK I have a little enhancement for the sphinx theme builder here that should let us do this safely: https://github.com/2i2c-org/sphinx-2i2c-theme/pull/18

@consideRatio would you be willing to give a quick look? If you're +1 then I think we can go ahead with making this switch in our docs. And we can test it out in:

- https://github.com/2i2c-org/team-compass/pull/667

**Comment by choldgraf** [comment]
Note that I also realized there was a hidden proposal above so I decided to make it more explicit for feedback:

- https://github.com/2i2c-org/team-compass/issues/705

**Comment by choldgraf** [comment]
### Two more PRs to move docs/infrastructure

- [ ] https://github.com/2i2c-org/docs/pull/183
- [ ] https://github.com/2i2c-org/infrastructure/pull/2508

**Comment by damianavila** [comment]
Since there is an automatic redirector from RTD, are you planning to revert https://github.com/2i2c-org/sphinx-2i2c-theme/pull/18, or do you want to keep it in case we do not use RTD?

**Comment by choldgraf** [comment]
I think we should try reverting it for one site and see if it works as expected, and if so revert the others. It's not top priority for me right now though so might be a little bit of time.
